### PR TITLE
Handling the Poisoning message problem using the Backout queue method.

### PR DESCRIPTION
**Pull request by rin0o1**
This PR implements the code for handling the poisoning message problem using the Backout Queue technique. This PR mainly affects the "*response*" script within the JMS, GO, Node.js, and Python projects. The keys and the env.json files might want to be deleted before merging. If the merge will succeed the README files must be updated.

### CODE REVIEW

**Review by chughts — CHANGES_REQUESTED** [review]
Copyright on all the changed files need to be changed to the form

````
* Copyright 2019, 2022 IBM Corp.
````

where the first date is the initial creation year, and the second date is the last change year.

As this needs to be updated, there are some minor syntax changes required.

File: JMS/dependency-reduced-pom.xml [hunk 1]
```diff
@@ -0,0 +1,87 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (chughts) [inline JMS/dependency-reduced-pom.xml:1]: Why are there 2 pom.xml files ?

File: JMS/target/maven-archiver/pom.properties [hunk 2]
```diff
@@ -0,0 +1,4 @@
+#Created by Apache Maven 3.8.6
```
Comment (chughts) [inline JMS/target/maven-archiver/pom.properties:1]: The target directory and its contents should not be part of this repository. Please remove from the pull request.

File: Node.js/basicresponse.js [hunk 3]
```diff
@@ -110,6 +111,46 @@ function getMessages(hConn, hObj) {
...
+      mqmd.ReplyToQ=MQDetails.BACKOUT_QUEUE;
+      debug_info("Redirecting the message to the bacout queue " + mqmd.ReplyToQ)
+      respondToRequest(hConn, hObj, msgObject, mqmd, true);
+      debug_info("Message redirected correctly")
+    }
+    catch(err){
+       ok=false 
+    }
+  }
+  else{
```
Comment (chughts) [inline Node.js/basicresponse.js:144]: space between else and {

File: Node.js/basicresponse.js [hunk 4]
```diff
@@ -110,6 +111,46 @@ function getMessages(hConn, hObj) {
...
+
+  var counter= mqmd.BackoutCount;
+  debug_info("------CURRENT BACKOUT COUNTER "+ counter);
+
+  if(counter >= MSG_TRESHOLD){
+   
+    try 
+    { 
+      mqmd.ReplyToQ=MQDetails.BACKOUT_QUEUE;
+      debug_info("Redirecting the message to the bacout queue " + mqmd.ReplyToQ)
```
Comment (chughts) [inline Node.js/basicresponse.js:136]: backout

File: Node.js/sampleresponse.js [hunk 5]
```diff
@@ -34,24 +38,95 @@ var MQBoilerPlate = require('./boilerplate');
...
+      debug_info('Reply Posted');
+      mqBoilerPlate.commit(callbackOnCommit);
+    })
+    .catch((err) => {
+      debug_warn('Error redirecting to the backout queue ');
+    });
+   
+    rollback=false;
+  }
+  else{
```
Comment (chughts) [inline Node.js/sampleresponse.js:95]: space between else and {

File: Python/basicput.py [hunk 6]
```diff
@@ -52,11 +52,13 @@ def connect():
         if MQDetails[EnvStore.KEY_REPOSITORY]:
```
Comment (chughts) [inline Python/basicput.py:52]: Other than an extra print, this file has not changed and should not be part or the pull request.

File: Python/basicrequest.py [hunk 7]
```diff
@@ -79,7 +79,7 @@ def getQueue():
         od = pymqi.OD()
         od.ObjectName = MQDetails[EnvStore.QUEUE_NAME]
         q.open(od, pymqi.CMQC.MQOO_OUTPUT)
-
+        print('Connected to queue ' + str(MQDetails[EnvStore.QUEUE_NAME]))
```
Comment (chughts) [inline Python/basicrequest.py:82]: use logger and not print.

File: Python/basicrequest.py [hunk 8]
```diff
@@ -94,13 +94,15 @@ def getDynamicQueue():
...
+        print(MQDetails[EnvStore.MODEL_QUEUE_NAME])
+        print(MQDetails[EnvStore.DYNAMIC_QUEUE_PREFIX])
         dyn_od.ObjectName = MQDetails[EnvStore.MODEL_QUEUE_NAME]
         dyn_od.DynamicQName = MQDetails[EnvStore.DYNAMIC_QUEUE_PREFIX]
 
         # Open the dynamic queue.
         dyn_input_open_options = pymqi.CMQC.MQOO_INPUT_EXCLUSIVE
         dyn_queue = pymqi.Queue(qmgr, dyn_od, dyn_input_open_options)
-
+        print("CREATED DYN QUEUE: " + str(dyn_queue))
```
Comment (chughts) [inline Python/basicrequest.py:105]: use logger

File: Python/basicresponse.py [hunk 9]
```diff
@@ -99,53 +100,103 @@ def getQueue(queueName, forInput):
...
+def rollback(qmgr , md, msg, backoutCounter):
+    # get the backout queue from the Environment --> fix this
+    BACKOUT_QUEUE = MQDetails[EnvStore.BACKOUT_QUEUE]
+   
+    ok=False 
+    #BACKOUT_QUEUE= 'DEV.QUEUE.2'
+    # if the backout counter is greater than 5
+    # handle possible poisoning message scenario
+    if(backoutCounter>=5):
+        print("POSIONING MESSAGE DETECTED! ")
```
Comment (chughts) [inline Python/basicresponse.py:178]: use logger and not print

File: Spring-JMS/target/classes/application.properties [hunk 10]
```diff
@@ -0,0 +1,134 @@
+# (c) Copyright IBM Corporation 2021
```
Comment (chughts) [inline Spring-JMS/target/classes/application.properties:1]: Why is this file included. Its a target from the spring library.

File: amqp-qpid/qpid-quarkus/target/classes/application.properties [hunk 11]
```diff
@@ -0,0 +1,20 @@
+# Configuration file
```
Comment (chughts) [inline amqp-qpid/qpid-quarkus/target/classes/application.properties:1]: should not be part of the pull request.

File: ccdt.json [hunk 12]
```diff
@@ -1,42 +0,0 @@
-{
```
Comment (chughts) [inline ccdt.json:1]: Why has the sample ccdt.json file been removed?

File: env-cloud.json [hunk 13]
```diff
@@ -0,0 +1,18 @@
+{
+  "MQ_ENDPOINTS": [{
+      "HOST": "test-a95a.qm.us-south.mq.appdomain.cloud",
+      "PORT": "32055",
+      "CHANNEL": "CLOUD.APP.SVRCONN",
+      "QMGR": "test",
+      "APP_USER": "credential2",
+      "APP_PASSWORD": "gOojnDA9y14IXhIH7gnKJA5AeTS_PFbz4FKda1KLuwSS",
```
Comment (chughts) [inline env-cloud.json:8]: Remember this is a public repository and all credentials are visible for ever, all it takes is to find the right commit. You should ensure that the credentials you have shared here are no-longer valid. It isn't sufficient to remove them as they can be refound.

File: qmgrcert.pem [hunk 14]
```diff
@@ -0,0 +1,94 @@
+-----BEGIN CERTIFICATE-----
```
Comment (chughts) [inline qmgrcert.pem:1]: please remove from the pull request.

**Review by chughts — CHANGES_REQUESTED** [review]
Copyright on all the changed files need to be changed to the form

````
* Copyright 2019, 2022 IBM Corp.
````

where the first date is the initial creation year, and the second date is the last change year.

As this needs to be updated, there are some minor syntax changes required.

A number of client key database are part of this pull request - please remove them

File: env-cloud.json (hunk 13 above)
Comment (rin0o1) [reply env-cloud.json:8]: The instance on the Cloud has been deleted.

File: Go/src/mqsamputils/connection.go [hunk 15]
```diff
@@ -65,20 +71,19 @@ func CreateConnection(index int) (ibmmq.MQQueueManager, error) {
 
 	// Allocate the MQCNO structure needed for the CONNX call.
 	cno := ibmmq.NewMQCNO()
-  env := getEndPoint(index)
+    env := getEndPoint(index)
```
Comment (rin0o1) [inline Go/src/mqsamputils/connection.go:74]: In my editor this line is aligned, whereas here it is not.

Comment (chughts) [reply Go/src/mqsamputils/connection.go:74]: I think there is a mix of spaces and tabs, you will need to run a code beautify.

**Review by chughts — CHANGES_REQUESTED** [review]
Nearly there, still need to fix that alignment problem in basic
response.go

**Review by chughts — APPROVED** [review]
